File: fraiseql-java/src/main/java/com/fraiseql/core/SchemaRegistry.java
package com.fraiseql.core;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Thread-safe registry for GraphQL types, queries, mutations, and subscriptions.
 * Singleton pattern ensures single schema instance across application.
 * Used by FraiseQL to track all registered types and root operations.
 */
public class SchemaRegistry {
    private static final SchemaRegistry INSTANCE = new SchemaRegistry();

    private final Map<String, GraphQLTypeInfo> types;
    private final Map<String, QueryInfo> queries;
    private final Map<String, MutationInfo> mutations;
    private final Map<String, SubscriptionInfo> subscriptions;
    private final Map<String, EnumInfo> enums;
    private final Map<String, InterfaceInfo> interfaces;
    private final Map<String, UnionInfo> unions;
    private final Map<String, InputTypeInfo> inputTypes;

    private SchemaRegistry() {
        this.types = new ConcurrentHashMap<>();
        this.queries = new ConcurrentHashMap<>();
        this.mutations = new ConcurrentHashMap<>();
        this.subscriptions = new ConcurrentHashMap<>();
        this.enums = new ConcurrentHashMap<>();
        this.interfaces = new ConcurrentHashMap<>();
        this.unions = new ConcurrentHashMap<>();
        this.inputTypes = new ConcurrentHashMap<>();
    }

    /**
     * Get the singleton instance of SchemaRegistry.
     *
     * @return the SchemaRegistry instance
     */
    public static SchemaRegistry getInstance() {
        return INSTANCE;
    }

    /**
     * Register a GraphQL type in the schema.
     *
     * @param typeName the GraphQL type name
     * @param typeClass the Java class annotated with @GraphQLType
     * @throws IllegalArgumentException if type is not annotated with @GraphQLType
     */
    public void registerType(String typeName, Class<?> typeClass) {
        if (!typeClass.isAnnotationPresent(GraphQLType.class)) {
            throw new IllegalArgumentException("Class " + typeClass.getName() + " must be annotated with @GraphQLType");
        }

        GraphQLType annotation = typeClass.getAnnotation(GraphQLType.class);
        String name = typeName;
        if (annotation.name() != null && !annotation.name().isEmpty()) {
            name = annotation.name();
        }

        Map<String, TypeConverter.GraphQLFieldInfo> fields = TypeConverter.extractFields(typeClass);

        GraphQLTypeInfo typeInfo = new GraphQLTypeInfo(
            name,
            typeClass,
            fields,
            annotation.description()
        );

        types.put(name, typeInfo);
    }

    /**
     * Register a query in the schema.
     *
     * @param queryName the query name
     * @param returnType the return type name
     * @param arguments the query arguments
     * @param description optional description
     */
    public void registerQuery(String queryName, String returnType, Map<String, String> arguments, String description) {
        QueryInfo queryInfo = new QueryInfo(queryName, returnType, arguments, description);
        queries.put(queryName, queryInfo);
    }

    /**
     * Register a mutation in the schema.
     *
     * @param mutationName the mutation name
     * @param returnType the return type name
     * @param arguments the mutation arguments
     * @param description optional description
     */
    public void registerMutation(String mutationName, String returnType, Map<String, String> arguments, String description) {
        MutationInfo mutationInfo = new MutationInfo(mutationName, returnType, arguments, description);
        mutations.put(mutationName, mutationInfo);
    }

    /**
     * Register a subscription in the schema.
     * Subscriptions in FraiseQL are compiled projections of database events.
     * They are sourced from LISTEN/NOTIFY or CDC, not resolver-based.
     *
     * @param subscriptionName the subscription name
     * @param entityType the entity type being subscribed to
     * @param arguments the subscription arguments (filters)
     * @param description optional description
     */
    public void registerSubscription(String subscriptionName, String entityType, Map<String, String> arguments, String description) {
        SubscriptionInfo subscriptionInfo = new SubscriptionInfo(subscriptionName, entityType, arguments, description, null, null);
        subscriptions.put(subscriptionName, subscriptionInfo);
    }

    /**
     * Register a subscription in the schema with topic and operation.
     * Subscriptions in FraiseQL are compiled projections of database events.
     * They are sourced from LISTEN/NOTIFY or CDC, not resolver-based.
     *
     * @param subscriptionName the subscription name
     * @param entityType the entity type being subscribed to
     * @param arguments the subscription arguments (filters)
     * @param description optional description
     * @param topic optional topic/channel name for filtering events
     * @param operation optional operation filter (CREATE, UPDATE, DELETE)
     */
    public void registerSubscription(String subscriptionName, String entityType, Map<String, String> arguments,
                                     String description, String topic, String operation) {
        SubscriptionInfo subscriptionInfo = new SubscriptionInfo(subscriptionName, entityType, arguments, description, topic, operation);
        subscriptions.put(subscriptionName, subscriptionInfo);
    }

    /**
     * Register an enum type in the schema.
     *
     * @param enumName the enum name
     * @param values the enum values (name -> value map)
     * @param description optional description
     */
    public void registerEnum(String enumName, Map<String, String> values, String description) {
        EnumInfo enumInfo = new EnumInfo(enumName, values, description);
        enums.put(enumName, enumInfo);
    }

    /**
     * Register an interface type in the schema.
     *
     * @param interfaceName the interface name
     * @param fields the interface fields
     * @param description optional description
     */
    public void registerInterface(String interfaceName, Map<String, TypeConverter.GraphQLFieldInfo> fields, String description) {
        InterfaceInfo interfaceInfo = new InterfaceInfo(interfaceName, fields, description);
        interfaces.put(interfaceName, interfaceInfo);
    }

    /**
     * Register a union type in the schema.
     *
     * @param unionName the union name
     * @param memberTypes the member type names
     * @param description optional description
     */
    public void registerUnion(String unionName, List<String> memberTypes, String description) {
        UnionInfo unionInfo = new UnionInfo(unionName, memberTypes, description);
        unions.put(unionName, unionInfo);
    }

    /**
     * Register an input type in the schema.
     *
     * @param inputName the input type name
     * @param fields the input fields
     * @param description optional description
     */
    public void registerInputType(String inputName, Map<String, TypeConverter.GraphQLFieldInfo> fields, String description) {
        InputTypeInfo inputInfo = new InputTypeInfo(inputName, fields, description);
        inputTypes.put(inputName, inputInfo);
    }

    /**
     * Get a registered type by name.
     *
     * @param typeName the type name
     * @return the GraphQLTypeInfo or empty Optional if not found
     */
    public Optional<GraphQLTypeInfo> getType(String typeName) {
        return Optional.ofNullable(types.get(typeName));
    }

    /**
     * Get a registered query by name.
     *
     * @param queryName the query name
     * @return the QueryInfo or empty Optional if not found
     */
    public Optional<QueryInfo> getQuery(String queryName) {
        return Optional.ofNullable(queries.get(queryName));
    }

    /**
     * Get a registered mutation by name.
     *
     * @param mutationName the mutation name
     * @return the MutationInfo or empty Optional if not found
     */
    public Optional<MutationInfo> getMutation(String mutationName) {
        return Optional.ofNullable(mutations.get(mutationName));
    }

    /**
     * Get a registered subscription by name.
     *
     * @param subscriptionName the subscription name
     * @return the SubscriptionInfo or empty Optional if not found
     */
    public Optional<SubscriptionInfo> getSubscription(String subscriptionName) {
        return Optional.ofNullable(subscriptions.get(subscriptionName));
    }

    /**
     * Get all registered types.
     *
     * @return unmodifiable map of type name to GraphQLTypeInfo
     */
    public Map<String, GraphQLTypeInfo> getAllTypes() {
        return Collections.unmodifiableMap(types);
    }

    /**
     * Get all registered queries.
     *
     * @return unmodifiable map of query name to QueryInfo
     */
    public Map<String, QueryInfo> getAllQueries() {
        return Collections.unmodifiableMap(queries);
    }

    /**
     * Get all registered mutations.
     *
     * @return unmodifiable map of mutation name to MutationInfo
     */
    public Map<String, MutationInfo> getAllMutations() {
        return Collections.unmodifiableMap(mutations);
    }

    /**
     * Get all registered subscriptions.
     *
     * @return unmodifiable map of subscription name to SubscriptionInfo
     */
    public Map<String, SubscriptionInfo> getAllSubscriptions() {
        return Collections.unmodifiableMap(subscriptions);
    }

    /**
     * Get an enum type by name.
     *
     * @param enumName the enum name
     * @return the EnumInfo or empty Optional if not found
     */
    public Optional<EnumInfo> getEnum(String enumName) {
        return Optional.ofNullable(enums.get(enumName));
    }

    /**
     * Get all registered enum types.
     *
     * @return unmodifiable map of enum name to EnumInfo
     */
    public Map<String, EnumInfo> getAllEnums() {
        return Collections.unmodifiableMap(enums);
    }

    /**
     * Get an interface type by name.
     *
     * @param interfaceName the interface name
     * @return the InterfaceInfo or empty Optional if not found
     */
    public Optional<InterfaceInfo> getInterface(String interfaceName) {
        return Optional.ofNullable(interfaces.get(interfaceName));
    }

    /**
     * Get all registered interface types.
     *
     * @return unmodifiable map of interface name to InterfaceInfo
     */
    public Map<String, InterfaceInfo> getAllInterfaces() {
        return Collections.unmodifiableMap(interfaces);
    }

    /**
     * Get a union type by name.
     *
     * @param unionName the union name
     * @return the UnionInfo or empty Optional if not found
     */
    public Optional<UnionInfo> getUnion(String unionName) {
        return Optional.ofNullable(unions.get(unionName));
    }

    /**
     * Get all registered union types.
     *
     * @return unmodifiable map of union name to UnionInfo
     */
    public Map<String, UnionInfo> getAllUnions() {
        return Collections.unmodifiableMap(unions);
    }

    /**
     * Get an input type by name.
     *
     * @param inputName the input type name
     * @return the InputTypeInfo or empty Optional if not found
     */
    public Optional<InputTypeInfo> getInputType(String inputName) {
        return Optional.ofNullable(inputTypes.get(inputName));
    }

    /**
     * Get all registered input types.
     *
     * @return unmodifiable map of input type name to InputTypeInfo
     */
    public Map<String, InputTypeInfo> getAllInputTypes() {
        return Collections.unmodifiableMap(inputTypes);
    }

    /**
     * Clear all registered types, queries, mutations, subscriptions, enums, interfaces, unions, and input types.
     * Useful for testing.
     */
    public void clear() {
        types.clear();
        queries.clear();
        mutations.clear();
        subscriptions.clear();
        enums.clear();
        interfaces.clear();
        unions.clear();
        inputTypes.clear();
    }

    /**
     * Information about a registered GraphQL type.
     */
    public static class GraphQLTypeInfo {
        public final String name;
        public final Class<?> javaClass;
        public final Map<String, TypeConverter.GraphQLFieldInfo> fields;
        public final String description;

        public GraphQLTypeInfo(String name, Class<?> javaClass, Map<String, TypeConverter.GraphQLFieldInfo> fields, String description) {
            this.name = name;
            this.javaClass = javaClass;
            this.fields = Collections.unmodifiableMap(new LinkedHashMap<>(fields));
            this.description = description;
        }

        @Override
        public String toString() {
            return "GraphQLTypeInfo{" +
                "name='" + name + '\'' +
                ", javaClass=" + javaClass.getSimpleName() +
                ", fields=" + fields.size() +
                '}';
        }
    }

    /**
     * Information about a registered GraphQL query.
     */
    public static class QueryInfo {
        public final String name;
        public final String returnType;
        public final Map<String, String> arguments;
        public final String description;

        public QueryInfo(String name, String returnType, Map<String, String> arguments, String description) {
            this.name = name;
            this.returnType = returnType;
            this.arguments = Collections.unmodifiableMap(new LinkedHashMap<>(arguments));
            this.description = description;
        }

        @Override
        public String toString() {
            return "QueryInfo{" +
                "name='" + name + '\'' +
                ", returnType='" + returnType + '\'' +
                ", arguments=" + arguments.size() +
                '}';
        }
    }

    /**
     * Information about a registered GraphQL mutation.
     */
    public static class MutationInfo {
        public final String name;
        public final String returnType;
        public final Map<String, String> arguments;
        public final String description;

        public MutationInfo(String name, String returnType, Map<String, String> arguments, String description) {
            this.name = name;
            this.returnType = returnType;
            this.arguments = Collections.unmodifiableMap(new LinkedHashMap<>(arguments));
            this.description = description;
        }

        @Override
        public String toString() {
            return "MutationInfo{" +
                "name='" + name + '\'' +
                ", returnType='" + returnType + '\'' +
                ", arguments=" + arguments.size() +
                '}';
        }
    }

    /**
     * Information about a registered GraphQL subscription.
     * Subscriptions in FraiseQL are compiled projections of database events.
     * They are sourced from LISTEN/NOTIFY or CDC, not resolver-based.
     */
    public static class SubscriptionInfo {
        public final String name;
        public final String entityType;
        public final Map<String, String> arguments;
        public final String description;
        public final String topic;
        public final String operation;

        public SubscriptionInfo(String name, String entityType, Map<String, String> arguments,
                                String description, String topic, String operation) {
            this.name = name;
            this.entityType = entityType;
            this.arguments = Collections.unmodifiableMap(new LinkedHashMap<>(arguments));
            this.description = description;
            this.topic = topic;
            this.operation = operation;
        }

        @Override
        public String toString() {
            return "SubscriptionInfo{" +
                "name='" + name + '\'' +
                ", entityType='" + entityType + '\'' +
                ", arguments=" + arguments.size() +
                (topic != null ? ", topic='" + topic + '\'' : "") +
                (operation != null ? ", operation='" + operation + '\'' : "") +
                '}';
        }
    }

    /**
     * Information about a registered GraphQL enum type.
     */
    public static class EnumInfo {
        public final String name;
        public final Map<String, String> values;
        public final String description;

        public EnumInfo(String name, Map<String, String> values, String description) {
            this.name = name;
            this.values = Collections.unmodifiableMap(new LinkedHashMap<>(values));
            this.description = description;
        }

        @Override
        public String toString() {
            return "EnumInfo{" +
                "name='" + name + '\'' +
                ", values=" + values.size() +
                '}';
        }
    }

    /**
     * Information about a registered GraphQL interface type.
     */
    public static class InterfaceInfo {
        public final String name;
        public final Map<String, TypeConverter.GraphQLFieldInfo> fields;
        public final String description;

        public InterfaceInfo(String name, Map<String, TypeConverter.GraphQLFieldInfo> fields, String description) {
            this.name = name;
            this.fields = Collections.unmodifiableMap(new LinkedHashMap<>(fields));
            this.description = description;
        }

        @Override
        public String toString() {
            return "InterfaceInfo{" +
                "name='" + name + '\'' +
                ", fields=" + fields.size() +
                '}';
        }
    }

    /**
     * Information about a registered GraphQL union type.
     */
    public static class UnionInfo {
        public final String name;
        public final List<String> memberTypes;
        public final String description;

        public UnionInfo(String name, List<String> memberTypes, String description) {
            this.name = name;
            this.memberTypes = Collections.unmodifiableList(new ArrayList<>(memberTypes));
            this.description = description;
        }

        @Override
        public String toString() {
            return "UnionInfo{" +
                "name='" + name + '\'' +
                ", memberTypes=" + memberTypes.size() +
                '}';
        }
    }

    /**
     * Information about a registered GraphQL input type.
     */
    public static class InputTypeInfo {
        public final String name;
        public final Map<String, TypeConverter.GraphQLFieldInfo> fields;
        public final String description;

        public InputTypeInfo(String name, Map<String, TypeConverter.GraphQLFieldInfo> fields, String description) {
            this.name = name;
            this.fields = Collections.unmodifiableMap(new LinkedHashMap<>(fields));
            this.description = description;
        }

        @Override
        public String toString() {
            return "InputTypeInfo{" +
                "name='" + name + '\'' +
                ", fields=" + fields.size() +
                '}';
        }
    }
}

File: sdks/official/fraiseql-php/src/SchemaRegistry.php
<?php

declare(strict_types=1);

namespace FraiseQL;

use FraiseQL\Attributes\GraphQLType;
use ReflectionClass;

/**
 * Central registry for GraphQL type definitions.
 *
 * The SchemaRegistry manages type registration, lookup, and introspection
 * for the entire GraphQL schema. It uses a singleton pattern suitable for
 * PHP's single-threaded execution model.
 *
 * Usage:
 * ```php
 * $registry = SchemaRegistry::getInstance();
 * $registry->register(User::class);
 * $userType = $registry->getType('User');
 * ```
 */
final class SchemaRegistry
{
    /** @var SchemaRegistry|null */
    private static ?self $instance = null;

    /** @var array<string, GraphQLType> Registered types by name */
    private array $types = [];

    /** @var array<string, string> Class name to GraphQL type name mapping */
    private array $classToTypeName = [];

    /** @var array<string, array<string, FieldDefinition>> Fields for each type */
    private array $typeFields = [];

    /** @var array<string, SubscriptionDefinition> Registered subscriptions */
    private array $subscriptions = [];

    /** @var array<string, QueryBuilder> Registered queries */
    private array $queries = [];

    /** @var array<string, MutationBuilder> Registered mutations */
    private array $mutations = [];

    /** @var array<string, array{name: string, fields: list<array{name: string, type: string, nullable: bool}>, description: string|null}> Registered input types */
    private array $inputTypes = [];

    /** @var array<string, array{sql_source: string|null, is_error: bool}> Extra type metadata */
    private array $typeMeta = [];

    private function __construct()
    {
    }

    /**
     * Get the singleton instance of the SchemaRegistry.
     *
     * @return self
     */
    public static function getInstance(): self
    {
        if (self::$instance === null) {
            self::$instance = new self();
        }

        return self::$instance;
    }

    /**
     * Register a GraphQL type from a PHP class.
     *
     * Extracts type definition from GraphQLType attribute and introspects
     * all properties marked with GraphQLField attributes.
     *
     * @param class-string $className The fully qualified class name
     * @return self Fluent interface
     *
     * @throws FraiseQLException If class doesn't have GraphQLType attribute
     */
    public function register(string $className): self
    {
        $reflection = new ReflectionClass($className);
        $attributes = $reflection->getAttributes(GraphQLType::class);

        if (empty($attributes)) {
            throw new FraiseQLException(
                "Class $className must have #[GraphQLType] attribute",
            );
        }

        /** @var GraphQLType $typeAttribute */
        $typeAttribute = $attributes[0]->newInstance();

        $typeName = $typeAttribute->name ?? $reflection->getShortName();

        // Register type
        $this->types[$typeName] = $typeAttribute;
        $this->classToTypeName[$className] = $typeName;

        // Register fields
        $fields = [];
        foreach ($reflection->getProperties() as $property) {
            $fields[$property->getName()] = $this->extractFieldDefinition($property, $typeName);
        }

        $this->typeFields[$typeName] = $fields;

        return $this;
    }

    /**
     * Get a registered type by name.
     *
     * @param string $typeName The GraphQL type name
     * @return GraphQLType|null The type definition, or null if not registered
     */
    public function getType(string $typeName): ?GraphQLType
    {
        return $this->types[$typeName] ?? null;
    }

    /**
     * Get all fields for a type.
     *
     * @param string $typeName The GraphQL type name
     * @return array<string, FieldDefinition> Array of field definitions indexed by field name
     */
    public function getTypeFields(string $typeName): array
    {
        return $this->typeFields[$typeName] ?? [];
    }

    /**
     * Get a specific field definition.
     *
     * @param string $typeName The GraphQL type name
     * @param string $fieldName The field name
     * @return FieldDefinition|null The field definition, or null if not found
     */
    public function getField(string $typeName, string $fieldName): ?FieldDefinition
    {
        return $this->typeFields[$typeName][$fieldName] ?? null;
    }

    /**
     * Check if a type is registered.
     *
     * @param string $typeName The GraphQL type name
     * @return bool
     */
    public function hasType(string $typeName): bool
    {
        return isset($this->types[$typeName]);
    }

    /**
     * Get all registered type names.
     *
     * @return array<string> Array of type names
     */
    public function getTypeNames(): array
    {
        return array_keys($this->types);
    }

    /**
     * Get the GraphQL type name for a PHP class.
     *
     * @param class-string $className The PHP class name
     * @return string|null The GraphQL type name, or null if class not registered
     */
    public function getTypeNameForClass(string $className): ?string
    {
        return $this->classToTypeName[$className] ?? null;
    }

    /**
     * Register a subscription.
     * Subscriptions in FraiseQL are compiled projections of database events.
     * They are sourced from LISTEN/NOTIFY or CDC, not resolver-based.
     *
     * @param SubscriptionDefinition $subscription The subscription to register
     * @return self Fluent interface
     */
    public function registerSubscription(SubscriptionDefinition $subscription): self
    {
        $this->subscriptions[$subscription->name] = $subscription;
        return $this;
    }

    /**
     * Get a registered subscription by name.
     *
     * @param string $name The subscription name
     * @return SubscriptionDefinition|null The subscription or null if not found
     */
    public function getSubscription(string $name): ?SubscriptionDefinition
    {
        return $this->subscriptions[$name] ?? null;
    }

    /**
     * Get all registered subscriptions.
     *
     * @return array<string, SubscriptionDefinition>
     */
    public function getAllSubscriptions(): array
    {
        return $this->subscriptions;
    }

    /**
     * Check if a subscription is registered.
     *
     * @param string $name The subscription name
     * @return bool
     */
    public function hasSubscription(string $name): bool
    {
        return isset($this->subscriptions[$name]);
    }


    /**
     * Register a query definition.
     *
     * @param QueryBuilder $query The query builder
     * @return self Fluent interface
     */
    public function registerQuery(QueryBuilder $query): self
    {
        $this->queries[$query->getName()] = $query;
        return $this;
    }

    /**
     * Get a registered query by name.
     *
     * @param string $name The query name
     * @return QueryBuilder|null
     */
    public function getQuery(string $name): ?QueryBuilder
    {
        return $this->queries[$name] ?? null;
    }

    /**
     * Get all registered queries.
     *
     * @return array<string, QueryBuilder>
     */
    public function getAllQueries(): array
    {
        return $this->queries;
    }

    /**
     * Register a mutation definition.
     *
     * @param MutationBuilder $mutation The mutation builder
     * @return self Fluent interface
     */
    public function registerMutation(MutationBuilder $mutation): self
    {
        $this->mutations[$mutation->getName()] = $mutation;
        return $this;
    }

    /**
     * Get a registered mutation by name.
     *
     * @param string $name The mutation name
     * @return MutationBuilder|null
     */
    public function getMutation(string $name): ?MutationBuilder
    {
        return $this->mutations[$name] ?? null;
    }

    /**
     * Get all registered mutations.
     *
     * @return array<string, MutationBuilder>
     */
    public function getAllMutations(): array
    {
        return $this->mutations;
    }

    /**
     * Register a GraphQL input object type.
     *
     * @param string $name Input type name (e.g. "CreateUserInput")
     * @param list<array{name: string, type: string, nullable: bool}> $fields Field definitions
     * @param string|null $description Optional input type description
     * @return self Fluent interface
     *
     * @throws FraiseQLException If an input type with this name is already registered
     */
    public function registerInputType(string $name, array $fields, ?string $description = null): self
    {
        if (isset($this->inputTypes[$name])) {
            throw new FraiseQLException(
                "Input type '{$name}' is already registered. Each name must be unique within a schema.",
            );
        }

        $this->inputTypes[$name] = [
            'name' => $name,
            'fields' => $fields,
            'description' => $description,
        ];

        return $this;
    }

    /**
     * Get all registered input types.
     *
     * @return array<string, array{name: string, fields: list<array{name: string, type: string, nullable: bool}>, description: string|null}>
     */
    public function getAllInputTypes(): array
    {
        return $this->inputTypes;
    }

    /** @var array<string, string> Base inject defaults */
    private array $injectDefaultsBase = [];

    /** @var array<string, string> Query inject defaults */
    private array $injectDefaultsQueries = [];

    /** @var array<string, string> Mutation inject defaults */
    private array $injectDefaultsMutations = [];

    /**
     * Set inject defaults from configuration.
     *
     * @param array<string, string> $base Base defaults for all operations
     * @param array<string, string> $queries Additional defaults for queries
     * @param array<string, string> $mutations Additional defaults for mutations
     * @return void
     */
    public function setInjectDefaults(array $base, array $queries, array $mutations): void
    {
        $this->injectDefaultsBase = $base;
        $this->injectDefaultsQueries = $queries;
        $this->injectDefaultsMutations = $mutations;
    }

    /**
     * Get inject defaults.
     *
     * @return array{base: array<string, string>, queries: array<string, string>, mutations: array<string, string>}
     */
    public function getInjectDefaults(): array
    {
        return [
            'base' => $this->injectDefaultsBase,
            'queries' => $this->injectDefaultsQueries,
            'mutations' => $this->injectDefaultsMutations,
        ];
    }

    /**
     * Get type metadata (sql_source, is_error) for a specific type.
     *
     * @param string $typeName The type name
     * @return array{sql_source: string|null, is_error: bool}|null
     */
    public function getTypeMeta(string $typeName): ?array
    {
        return $this->typeMeta[$typeName] ?? null;
    }

    /**
     * Set type metadata (sql_source, is_error) for a specific type.
     *
     * @param string $typeName The type name
     * @param array{sql_source: string|null, is_error: bool} $meta The metadata
     * @return void
     */
    public function setTypeMeta(string $typeName, array $meta): void
    {
        $this->typeMeta[$typeName] = $meta;
    }

    /**
     * Clear all registered types (useful for testing).
     *
     * @return self Fluent interface
     */
    public function clear(): self
    {
        $this->types = [];
        $this->classToTypeName = [];
        $this->typeFields = [];
        $this->subscriptions = [];
        $this->queries = [];
        $this->mutations = [];
        $this->inputTypes = [];
        $this->typeMeta = [];
        $this->injectDefaultsBase = [];
        $this->injectDefaultsQueries = [];
        $this->injectDefaultsMutations = [];

        return $this;
    }

    /**
     * Extract field definition from a ReflectionProperty.
     *
     * @param \ReflectionProperty $property The property to extract from
     * @param string $typeName The parent type name
     * @return FieldDefinition The field definition
     */
    private function extractFieldDefinition(
        \ReflectionProperty $property,
        string $typeName,
    ): FieldDefinition {
        $typeInfo = TypeConverter::fromReflectionProperty($property);

        return new FieldDefinition(
            name: $property->getName(),
            type: $typeInfo->graphQLType,
            nullable: $typeInfo->isNullable,
            isList: $typeInfo->isList,
            description: $typeInfo->description,
            phpType: $typeInfo->phpType,
            customResolver: $typeInfo->customResolver,
            parentType: $typeName,
        );
    }
}
